runtime/parachains/src/disputes.rs

```diff
@@ -0,0 +1,1963 @@
+		if !is_local {
+			let mut spam_slots: Vec<u32> = SpamSlots::<T>::get(&set.session)
+				.unwrap_or_else(|| vec![0; n_validators]);
+
+			for (validator_index, spam_slot_change) in summary.spam_slot_changes {
+				let spam_slot = spam_slots.get_mut(validator_index.0 as usize)
+					.expect("index is in-bounds, as checked above; qed");
+
+				match spam_slot_change {
+					SpamSlotChange::Inc => {
+						ensure!(
+							*spam_slot < config.dispute_max_spam_slots,
+							Error::<T>::PotentialSpam,
+						);
+
+						*spam_slot += 1;
+					}
+					SpamSlotChange::Dec => {
+						*spam_slot = spam_slot.saturating_sub(1);
+					}
+				}
+			}
+
+			SpamSlots::<T>::insert(&set.session, spam_slots);
+		}
+
+		if fresh {
+			Self::deposit_event(Event::DisputeInitiated(
+				set.candidate_hash,
+				if is_local { DisputeLocation::Local } else { DisputeLocation::Remote },
+			));
+		}
+
+		{
+			if summary.new_flags.contains(DisputeStateFlags::FOR_SUPERMAJORITY) {
+				Self::deposit_event(Event::DisputeConcluded(
+					set.candidate_hash,
+					DisputeResult::Valid,
+				));
+			}
+
+			// It is possible, although unexpected, for a dispute to conclude twice.
+			// This would require f+1 validators to vote in both directions.
+			// A dispute cannot conclude more than once in each direction.
+
+			if summary.new_flags.contains(DisputeStateFlags::AGAINST_SUPERMAJORITY) {
+				Self::deposit_event(Event::DisputeConcluded(
+					set.candidate_hash,
+					DisputeResult::Invalid,
+				));
+			}
+		}
+
+		// Reward statements.
+		T::RewardValidators::reward_dispute_statement(
+			set.session,
+			summary.new_participants.iter_ones().map(|i| ValidatorIndex(i as _)),
+		);
+
+		// Slash participants on a losing side.
+		{
+			// a valid candidate, according to 2/3. Punish those on the 'against' side.
+			T::PunishValidators::punish_against_valid(
+				set.session,
+				summary.slash_against,
+			);
+
+			// an invalid candidate, according to 2/3. Punish those on the 'for' side.
+			T::PunishValidators::punish_for_invalid(
+				set.session,
+				summary.slash_for,
+			);
+		}
+
+		<Disputes<T>>::insert(&set.session, &set.candidate_hash, &summary.state);
+
+		// Freeze if just concluded against some local candidate
+		if summary.new_flags.contains(DisputeStateFlags::AGAINST_SUPERMAJORITY) {
+			if let Some(revert_to) = <Included<T>>::get(&set.session, &set.candidate_hash) {
+				Self::revert_and_freeze(revert_to);
+			}
+		}
+
+		Ok(fresh)
+	}
+
+	#[allow(unused)]
+	pub(crate) fn disputes() -> Vec<(SessionIndex, CandidateHash, DisputeState<T::BlockNumber>)> {
+		<Disputes<T>>::iter().collect()
+	}
+
+	pub(crate) fn note_included(session: SessionIndex, candidate_hash: CandidateHash, included_in: T::BlockNumber) {
+		if included_in.is_zero() { return }
+
+		let revert_to = included_in - One::one();
+
+		<Included<T>>::insert(&session, &candidate_hash, revert_to);
+
+		// If we just included a block locally which has a live dispute, decrement spam slots
```

Comment:
In principle the client shouldn't be doing this in the first place right?

Reply:
It shouldn't, but it might happen especially at the exact point of reversion. If there's a bad candidate C which is included at a block B+1', and then we revert back to B, there's the possibility that the transplanted dispute and the candidate both get included. That leads to the new chain also needing to be abandoned, of course.

What I want honest nodes to do is avoid importing availability bitfields on stale leaves, so this scenario won't happen unless the node authoring the first block after a revert is malicious. And this code is to defend against such nodes